ios-base/Networking/Services/APIClient.swift

```diff
@@ -0,0 +1,341 @@
+                     paramsEncoding: ParameterEncoding? = nil,
+                     success: @escaping SuccessCallback,
+                     failure: @escaping FailureCallback) {
+    let encoding = paramsEncoding ?? defaultEncoding(forMethod: method)
+    let headers = APIClient.getHeaders()
+    let requestConvertible = BaseURLRequestConvertible(
+      path: url,
+      method: method,
+      encoding: encoding,
+      params: params,
+      headers: headers
+    )
+    
+    let request = AF.request(requestConvertible)
+    
+    request.responseJSON(
+      completionHandler: { result in
+        validateResult(result: result, success: success, failure: failure)
+    })
+  }
+  
+  //Handle rails-API-base errors if any
+  class func handleCustomError(_ code: Int?, dictionary: [String: Any]) -> NSError? {
+    if
+      let messageDict = dictionary["errors"] as? [String: [String]],
+      let firstKey = messageDict.keys.first
+    {
+      let errorsList = messageDict[firstKey]
+      return NSError(
+        domain: "\(firstKey) \(errorsList?.first ?? "")",
+        code: code ?? 500,
+        userInfo: nil
+      )
+    } else if let error = dictionary["error"] as? String {
+      return NSError(domain: error, code: code ?? 500, userInfo: nil)
+    } else if
+      let errors = dictionary["errors"] as? [String: Any],
+      let firstKey = errors.keys.first
+    {
+      let errorDesc = errors[firstKey] ?? ""
+      return NSError(
+        domain: "\(firstKey) " + "\(errorDesc)",
+        code: code ?? 500,
+        userInfo: nil
+      )
+    } else if dictionary["errors"] != nil || dictionary["error"] != nil {
+      return NSError(
+        domain: "Something went wrong. Try again later.",
+        code: code ?? 500,
+        userInfo: nil
+      )
+    }
+    return nil
+  }
+  
+  fileprivate class func validateResult(
+    result: AFDataResponse<Any>,
+    success: @escaping SuccessCallback,
+    failure: @escaping FailureCallback
+  ) {
+    let defaultError = App.error(
+      domain: .generic,
+      localizedDescription: "Error parsing response".localized
+    )
+  
+    guard let response = result.response else {
+      failure(defaultError)
+      return
+    }
+    
+    guard !validateEmptyResponse(
+      response: response,
+      data: result.data,
+      success: success,
+      failure: failure
+    ) else { return }
+    
+    guard let data = result.data else {
+      failure(defaultError)
+      return
+    }
+    
+    validateSerializationErrors(
+      response: response,
+      error: result.error,
+      data: data,
+      success: success,
+      failure: failure
+    )
+  }
+  
+  fileprivate class func validateEmptyResponse(
+    response: HTTPURLResponse,
+    data: Data?,
+    success: @escaping SuccessCallback,
+    failure: @escaping FailureCallback
+  ) -> Bool {
+    let defaultError = App.error(
+      domain: .generic,
+      localizedDescription: "Error parsing response".localized
```

Comment:
maybe a more suitable error for this case would be "Unexpected empty response" or something like that.